# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/test/test_dynamic_change_causing_reflow.html
# Commit: b2727a598452
# Full Hash: b2727a598452fb942e0611df5f931d4df0efaceb
# Author: Frédéric Wang <fwang@igalia.com>
# Date: 2024-01-18 04:47:51
# Regressor Bug: 1765615
# File Overlap Count: 1
# Description:
#   Bug 1765615 - Handle most changes to CSS `contain` and `content-visibility` without needing to reconstruct frames. r=layout-reviewers,emilio
#   
#   Right now, we reconstruct frames in response to a change in the CSS
#   `contain` property or `content-visibility`. This patch tries to optimize
#   this a bit:
# ==============================================================================

diff -r 5608664350d8 -r b2727a598452 layout/style/test/test_dynamic_change_causing_reflow.html
--- a/layout/style/test/test_dynamic_change_causing_reflow.html	Wed Jan 17 07:22:29 2024 +0000
+++ b/layout/style/test/test_dynamic_change_causing_reflow.html	Wed Jan 17 08:22:48 2024 +0000
@@ -34,6 +34,9 @@
   #tableCellWithAbsPosChild {
     display: table-cell;
   }
+  .blockmargin {
+    margin: 25px 0;
+  }
 </style>
 <div id="display">
   <div id="content">
@@ -76,6 +79,24 @@
   <div id="tableCellWithAbsPosChild">
     <div style="position: absolute"></div>
   </div>
+  <div id="containNode">
+    <div></div>
+  </div>
+  <div id="containNodeWithAbsPosChild">
+    <div style="position: absolute"></div>
+  </div>
+  <div id="containNodeWithFixedPosChild">
+    <div style="position: fixed"></div>
+  </div>
+  <div id="containNodeWithFloatingChild">
+    <div style="float: left"></div>
+  </div>
+  <div id="containNodeWithMarginCollapsing" class="blockmargin">
+    <div class="blockmargin"></div>
+  </div>
+  <div id="contentVisibilityNode">
+    <div></div>
+  </div>
 </div>
 <pre id="test">
 <script>
@@ -633,6 +654,284 @@
     expectConstruction: false,
     expectReflow: false,
   },
+  // Style containment affects counters so we need to re-frame.
+  // For other containments, we only need to reflow.
+  {
+    elem: containNode,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: style",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: style",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: paint",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: paint",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: layout",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: layout",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: size",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNode,
+    beforeStyle: "contain: size",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  // paint/layout containment boxes establish an absolute positioning
+  // containing block, so need a re-frame to handle abs/fixed positioned boxes.
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: paint",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: paint",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: layout",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: layout",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: size",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithAbsPosChild,
+    beforeStyle: "contain: size",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: paint",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: paint",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: layout",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: layout",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: size",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFixedPosChild,
+    beforeStyle: "contain: size",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  // paint/layout containment boxes establish an independent formatting context,
+  // so need a re-frame to handle floated boxes.
+  // FIXME(bug 1874826): Try and avoid a full construction.
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: paint",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: paint",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: layout",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: layout",
+    afterStyle: "contain: none",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: size",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithFloatingChild,
+    beforeStyle: "contain: size",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  // However, a reflow is enough to update margin collapsing.
+  {
+    elem: containNodeWithMarginCollapsing,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: paint",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithMarginCollapsing,
+    beforeStyle: "contain: paint",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithMarginCollapsing,
+    beforeStyle: "contain: none",
+    afterStyle: "contain: layout",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: containNodeWithMarginCollapsing,
+    beforeStyle: "contain: layout",
+    afterStyle: "contain: none",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  // content-visibility: auto/hidden implies style containment contrary to
+  // content-visibility: visible, so  we generally need a re-frame (see above)
+  // when going from one case to the other.
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: visible",
+    afterStyle: "content-visibility: hidden",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: hidden",
+    afterStyle: "content-visibility: visible",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: visible",
+    afterStyle: "content-visibility: auto",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: auto",
+    afterStyle: "content-visibility: visible",
+    expectConstruction: true,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: hidden",
+    afterStyle: "content-visibility: auto",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: auto",
+    afterStyle: "content-visibility: hidden",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  // However that's not the case if we force style containment explicitly.
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: visible; contain: style",
+    afterStyle: "content-visibility: hidden; contain: style",
+    expectConstruction: false,
+    expectReflow: true,
+  },
+  {
+    elem: contentVisibilityNode,
+    beforeStyle: "content-visibility: hidden; contain: style",
+    afterStyle: "content-visibility: visible; contain: style",
+    expectConstruction: false,
+    expectReflow: true,
+  },
 ];
 
 // Helper function to let us call either "is" or "isnot" & assemble
